pkg/providers/instance/aksmachineinstance.go

```diff
@@ -0,0 +1,580 @@
+	// Note that 'Get' also satisfies cloudprovider.Delete contract expectation (from v1.3.0)
+	// of returning cloudprovider.NewNodeClaimNotFoundError if the instance is already deleted
+	aksMachine, err := p.Get(ctx, aksMachineName)
+	if err != nil {
+		return err
+	}
+	if IsAKSMachineDeleting(aksMachine) {
+		return nil
+	}
+
+	err = p.deleteMachine(ctx, aksMachineName)
+	if err != nil {
+		if IsARMNotFound(err) {
+			return corecloudprovider.NewNodeClaimNotFoundError(fmt.Errorf("failed to delete AKS machine instance, %w", err))
+		}
+		return fmt.Errorf("failed to delete AKS machine instance, %w", err)
+	}
+	return nil
+}
+
+func (p *DefaultAKSMachineProvider) GetMachinesPoolLocation() string {
+	return p.aksMachinesPoolLocation
+}
+
+func (p *DefaultAKSMachineProvider) deleteMachine(ctx context.Context, aksMachineName string) error {
+	log.FromContext(ctx).V(1).Info("deleting AKS machine", "aksMachineName", aksMachineName)
+
+	// Suggestion: we could utilize this batch capability to optimize performance
+	aksMachines := armcontainerservice.AgentPoolDeleteMachinesParameter{
+		MachineNames: []*string{&aksMachineName},
+	}
+
+	poller, err := p.azClient.agentPoolsClient.BeginDeleteMachines(ctx, p.clusterResourceGroup, p.clusterName, p.aksMachinesPoolName, aksMachines, nil)
+	if err != nil {
+		return fmt.Errorf("failed to begin delete for AKS machine %q: %w", aksMachineName, err)
+	}
+
+	_, err = poller.PollUntilDone(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("failed to delete AKS machine %q: %w", aksMachineName, err)
+	}
+
+	log.FromContext(ctx).V(1).Info("successfully deleted AKS machine", "aksMachineName", aksMachineName)
+	return nil
+}
+
+// beginCreateMachine starts the creation of an AKS machine instance.
+// The returned AKSMachinePromise must be called to gather any errors
+// that are retrieved during async provisioning, as well as to complete the provisioning process.
+// nolint: gocyclo
+func (p *DefaultAKSMachineProvider) beginCreateMachine(
+	ctx context.Context,
+	nodeClass *v1beta1.AKSNodeClass,
+	nodeClaim *karpv1.NodeClaim,
+	instanceTypes []*corecloudprovider.InstanceType,
+	aksMachineName string,
+) (*AKSMachinePromise, error) {
+	// Reuse the existing AKS machine if it exists, and skip the creation.
+	// Note: currently, we do not support different offerings requirements for the NodeClaim with the same name that attempted creation recently. The same applies with VM-based provisioning.
+	// This supported case is often a result of Karpenter restarting while the node is being created, in which the NodeClaim to create stays the same after restart.
+	resp, err := p.azClient.aksMachinesClient.Get(ctx, p.clusterResourceGroup, p.clusterName, p.aksMachinesPoolName, aksMachineName, nil)
+	if err == nil {
+		// Existing AKS machine found, reuse it.
+
+		// Reconstruct properties from existing AKS machine instance.
+		if len(resp.Machine.Zones) == 0 || resp.Machine.Zones[0] == nil {
```

Comment:
This seems strange/wrong to me possibly?

What if machine exists but wrong configuration? Returning success for beginCreate seems incorrect then? Plus API should be idempotent if we PUT w/ same configuration again, so why do we need to do this GET check at all?